tests/python/test_multidevice.py

```diff
@@ -332,35 +332,49 @@ def definition(self) -> None:
         T78 = self.ops.linear(T77, self.mha_linear0_weight, self.mha_linear0_bias)
         T91 = self.ops.slice(
             T78,
-            start_indices=[0, 0, 0],
-            end_indices=[b, s, e],
-            strides=[1, 1, 1],
+            start_indices=[0, 0, 0, 0],
+            end_indices=[d, b, s, e // d],
         )
         T104 = self.ops.slice(
             T78,
-            start_indices=[0, 0, e],
-            end_indices=[b, s, e * 2],
-            strides=[1, 1, 1],
+            start_indices=[0, 0, 0, e // d],
+            end_indices=[d, b, s, e * 2 // d],
         )
         T117 = self.ops.slice(
             T78,
-            start_indices=[0, 0, e * 2],
-            end_indices=[b, s, e * 3],
-            strides=[1, 1, 1],
+            start_indices=[0, 0, 0, e * 2 // d],
+            end_indices=[d, b, s, e * 3 // d],
         )
-        T123 = self.ops.reshape(T104, new_shape=[b, s, h, e // h])
-        T124 = self.ops.permute(T123, dims=[0, 2, 1, 3])
-        T130 = self.ops.reshape(T91, new_shape=[b, s, h, e // h])
-        T131 = self.ops.permute(T130, dims=[0, 2, 1, 3])
-        T137 = self.ops.reshape(T117, new_shape=[b, s, h, e // h])
-        T138 = self.ops.permute(T137, dims=[0, 2, 1, 3])
+        T123 = self.ops.reshape(T104, new_shape=[d, b, s, h // d, e // h])
+        T124 = self.ops.permute(T123, dims=[0, 1, 3, 2, 4])
+        T130 = self.ops.reshape(T91, new_shape=[d, b, s, h // d, e // h])
+        T131 = self.ops.permute(T130, dims=[0, 1, 3, 2, 4])
+        T137 = self.ops.reshape(T117, new_shape=[d, b, s, h // d, e // h])
+        T138 = self.ops.permute(T137, dims=[0, 1, 3, 2, 4])
         S139 = self.define_scalar(0.100000, dtype=DataType.Double)
         S140 = self.define_scalar(True, dtype=DataType.Bool)
         T141, T142, T143, T144 = self.ops.sdpfa_fwd(T131, T124, T138, S139, S140, None)
-        T145 = self.ops.permute(T141, dims=[0, 2, 1, 3])
-        T146 = self.ops.stride_order(T145, stride_order=[3, 2, 1, 0])
-        T151 = self.ops.reshape(T146, new_shape=[b, s, e])
-        T152 = self.ops.linear(T151, self.mha_linear1_weight, self.mha_linear1_bias)
+        T145 = self.ops.permute(T141, dims=[0, 1, 3, 2, 4])
+        T146 = self.ops.stride_order(T145, stride_order=[4, 3, 2, 1, 0])
+        T151 = self.ops.reshape(T146, new_shape=[d, b, s, e // d])
+        # TODO(#3125): nvFuser is missing an API to construct a sharded linear
+        # like this. Therefore, I decomposed it by hand.
+        # T152 = self.ops.linear(T151, self.mha_linear1_weight, self.mha_linear1_bias)
+        #                   [d,b,s,e/d]        [d,e,e/d]                 [e]
+        T152_local_matmul = self.ops.matmul(
+            T151,
+            self.ops.broadcast_in_dim(
+                self.ops.permute(self.mha_linear1_weight, [0, 2, 1]),
+                [d, 1, e // d, e],
+                [0, 2, 3],
+            ),
+        )
+        T152_matmul = self.ops.sum(T152_local_matmul, [0])  # allreduce
```

Comment:
I wonder what would happen currently if we do not decompose the matmul and the allreduce...

Reply:
The first thing that'll break is that `linear` will produce a wrong shape. linear, as is implemented today, will output a tensor of rank `input_rank + weight_rank - 2 = 5`. However, we want the shape to be `[d,b,s,e]` and thus 4D.